build: Enhance Ccache performance across worktrees and build trees

hebasto opened the pull request:
To maximize performance, it is essential to follow the [recommendations](https://github.com/bitcoin/bitcoin/blob/master/doc/productivity.md#cache-compilations-with-ccache) in the Productivity Notes: set the `base_dir` Ccache configuration option or use the `CCACHE_BASEDIR` environment variable:
```
$ export CCACHE_BASEDIR=$HOME
```

Here are examples of usage on Ubuntu 24.04:

1. Across different worktrees:
```
$ export CCACHE_DIR=$(mktemp -d)
$ git worktree add ../wt_1 4d3f360e2af94f4ed46dc0943196d548da82003e
$ pushd ../wt_1
$ cmake -B build -DWITH_CCACHE=ON
$ cmake --build build -t bitcoind -j $(nproc)
$ popd
$ git worktree add ../wt_2 4d3f360e2af94f4ed46dc0943196d548da82003e
$ pushd ../wt_2
$ cmake -B build -DWITH_CCACHE=ON
$ ccache --zero-stats
$ cmake --build build -t bitcoind -j $(nproc)
$ popd
$ ccache --show-stats
Cacheable calls:    302 / 302 (100.0%)
  Hits:             302 / 302 (100.0%)
    Direct:         302 / 302 (100.0%)
    Preprocessed:     0 / 302 ( 0.00%)
  Misses:             0 / 302 ( 0.00%)
Local storage:
  Cache size (GiB): 0.2 / 5.0 ( 3.31%)
  Hits:             302 / 302 (100.0%)
  Misses:             0 / 302 ( 0.00%)
```

2. Across different build trees:
```
$ export CCACHE_DIR=$(mktemp -d)
$ cmake -B build_1 -DWITH_CCACHE=ON
$ cmake --build build_1 -t bitcoind -j $(nproc)
$ cmake -B build_2 -DWITH_CCACHE=ON
$ ccache --zero-stats
$ cmake --build build_2 -t bitcoind -j $(nproc)
$ ccache --show-stats
Cacheable calls:    302 / 302 (100.0%)
  Hits:             302 / 302 (100.0%)
    Direct:         302 / 302 (100.0%)
    Preprocessed:     0 / 302 ( 0.00%)
  Misses:             0 / 302 ( 0.00%)
Local storage:
  Cache size (GiB): 0.2 / 5.0 ( 3.31%)
  Hits:             302 / 302 (100.0%)
  Misses:             0 / 302 ( 0.00%)
```

This PR addresses this [comment](https://github.com/bitcoin/bitcoin/pull/30811#issuecomment-2331096969):
> However, ccache does _not_ hit across two different build dirs, compiling the same commit.

---

Fixes https://github.com/bitcoin/bitcoin/issues/31771.

Form https://github.com/bitcoin/bitcoin/actions/runs/13215680798/job/36894673006:
```
...
Treat compiler warnings as errors ..... ON
Use ccache for compiling .............. Built-in ccache support for the 'Visual Studio 17 2022' generator is not available.


-- Configuring done (300.1s)
-- Generating done (1.2s)
-- Build files have been written to: D:/a/bitcoin/bitcoin/build
```

fanquake commented:
> 100% cache hit rate is expected.

I tested this, and got about 90%:
```bash
ccache -C
ccache --zero-stats

cmake -B some_build_dir -DWITH_CCACHE=ON
cmake --build some_build_dir -j17

ccache --show-stats
Cacheable calls:    445 /  445 (100.0%)
  Hits:               0 /  445 ( 0.00%)
    Direct:           0
    Preprocessed:     0
  Misses:           445 /  445 (100.0%)
Local storage:
  Cache size (GiB): 0.2 / 30.0 ( 0.75%)
  Hits:               0 /  445 ( 0.00%)
  Misses:           445 /  445 (100.0%)

ccache --zero-stats

cmake -B /tmp/some_other_build_dir -DWITH_CCACHE=ON
cmake --build /tmp/some_other_build_dir -j17

ccache --show-stats
Cacheable calls:    445 /  445 (100.0%)
  Hits:             394 /  445 (88.54%)
    Direct:         394 /  394 (100.0%)
    Preprocessed:     0 /  394 ( 0.00%)
  Misses:            51 /  445 (11.46%)
Local storage:
  Cache size (GiB): 0.2 / 30.0 ( 0.76%)
  Hits:             394 /  445 (88.54%)
  Misses:            51 /  445 (11.46%)
```

hebasto commented on cmake/ccache.cmake:
```diff
@@ -2,7 +2,7 @@
 # file COPYING or https://opensource.org/license/mit/.
 
-if(NOT MSVC)
+if(CMAKE_GENERATOR MATCHES "Ninja|Makefiles")
```
When building across different build directories, using the `base_dir` option helps work around absolute paths in `-I` compiler options, increasing the `ccache` cache hit rate.

Maybe, the word "effectiveness" is more appropriate, rather than "performance".

The change in the condition is not directly related to improving the hit rate. It enhances code correctness, as the method using of the `CMAKE_{C,CXX}_COMPILER_LAUNCHER` variables is applicable only to the mentioned generators.

ryanofsky replied:
In commit "build: Improve `ccache` performance for different build directories" (66b73bbf375ff96bf4a77539c9cf3fe52fbc15d5)

Would be good to add a comment explaining `if(CMAKE_GENERATOR MATCHES "Ninja|Makefiles")` saying that COMPILER_LAUNCHER variable only works for these backends not other backends. Otherwise the check seems completely arbitrary and doesn't make sense.

hebasto replied:
> Would be good to add a comment explaining `if(CMAKE_GENERATOR MATCHES "Ninja|Makefiles")` saying that COMPILER_LAUNCHER variable only works for these backends not other backends. Otherwise the check seems arbitrary and doesn't seem to make sense.

Thanks! Done.

ryanofsky commented on cmake/ccache.cmake:
```diff
@@ -17,8 +17,11 @@ if(NOT MSVC)
-      list(APPEND CMAKE_CXX_COMPILER_LAUNCHER ${CCACHE_EXECUTABLE})
+      foreach(lang IN ITEMS C CXX OBJCXX)
+        set(CMAKE_${lang}_COMPILER_LAUNCHER
+          ${CCACHE_EXECUTABLE} base_dir=${CMAKE_BINARY_DIR}
```
In commit "build: Improve `ccache` performance for different build directories" (66b73bbf375ff96bf4a77539c9cf3fe52fbc15d5)

Reading https://ccache.dev/manual/latest.html#_configuration_options it seems like it would make sense in most cases to set base_dir to CMAKE_SOURCE_DIR not CMAKE_BINARY_DIR.

For example if CMAKE_SOURCE_DIR is /home/bitcoin and CMAKE_BINARY_DIR is /home/bitcoin/build, it sounds like ccmake will *not* rewrite the include path `-I/home/bitcoin/src` to `-I../src` for greater cache reuse if base_dir=CMAKE_BINARY_DIR is used, because /home/bitcoin/src is not a subdirectory of /home/bitcoin/build

But it would rewrite the include path if CMAKE_SOURCE_DIR were used, because /home/bitcoin/src is a subdirectory of /home/bitcoin/src.

ryanofsky replied:
> `CCACHE_BASEDIR=${CMAKE_BINARY_DIR}` is effective when building in different build directories from the same source directory.
> 
> `CCACHE_BASEDIR=${CMAKE_SOURCE_DIR}` works for building from different source directories.
> 
> We need to choose one of these two options.

Thanks, that clarifies a lot. I only do the second not the first, because I use git worktrees, and I assumed this PR was trying to make git worktrees and multiple checkouts work better. Maybe people who work on build systems are likely to have many different build directories inside the same source directory, but that would seem like an unusual thing for most other developers. Especially to have multiple build directories building with the same compilation options, because if compile options are different ccache command lines wouldn't match anyway.

Is there a real use-case for having multiple different build directories with the same compilation options in the same source directory? And optimizing to have cache hits in that case?

Also maybe I need to think about it more, but I'm not sure why choosing CMAKE_BINARY_DIR would be better than CMAKE_SOURCE_DIR in either case as long a CMAKE_BINARY_DIR is a subdirectory of CMAKE_SOURCE_DIR. Are we trying to optimize for cases where CMAKE_BINARY_DIR is not a subdirectory?

hebasto replied:
> Also maybe I need to think about it more, but I'm not sure why choosing CMAKE_BINARY_DIR would be better than CMAKE_SOURCE_DIR in either case as long a CMAKE_BINARY_DIR is a subdirectory of CMAKE_SOURCE_DIR. Are we trying to optimize for cases where CMAKE_BINARY_DIR is not a subdirectory?

I don't think it is related to whether `CMAKE_BINARY_DIR` is a subdirectory of `CMAKE_SOURCE_DIR`.

Due to https://github.com/bitcoin/bitcoin/blob/07c7c96022dd325be1cd3b353d575eb6a5593f55/src/CMakeLists.txt#L9
all objects are compiled with the `-I/absolute/path/to/build/dir` and `-I/absolute/path/to/source/dir` flags, and any difference in either will trigger a cache miss.

The "[Cache compilations with `ccache`](https://github.com/bitcoin/bitcoin/blob/master/doc/productivity.md#cache-compilations-with-ccache)" section of the Productivity Notes still applicable for this PR branch except for the last commit, which was mostly motivated by https://github.com/bitcoin/bitcoin/pull/30811#issuecomment-2331096969.

ryanofsky replied:
> I don't think it is related to whether `CMAKE_BINARY_DIR` is a subdirectory of `CMAKE_SOURCE_DIR` or not.

Ccache documentation describes behavior of base_dir: "ccache will rewrite absolute paths into paths relative to the current working directory, but only absolute paths that begin with **base_dir**"

So if CMAKE_BINARY_DIR is a subdirectory of CMAKE_SOURCE_DIR, and base_dir is specified as CMAKE_SOURCE_DIR, then that should be a strict improvement because it will allow cache hits across multiple source directories, and multiple build directories, as long as build directories are not outside the source directories.

hebasto replied:
> So if CMAKE_BINARY_DIR is a subdirectory of CMAKE_SOURCE_DIR, and base_dir is specified as CMAKE_SOURCE_DIR, then that should be a strict improvement because it will allow cache hits across multiple source directories, and multiple build directories, as long as build directories are not outside the source directories.

For the latter case, using different build directories will result in cache misses anyway, right?

Anyway, I agree with @maflcko and @ryanofsky that using git worktrees should be prioritised.

ryanofsky replied:
> > So if CMAKE_BINARY_DIR is a subdirectory of CMAKE_SOURCE_DIR, and base_dir is specified as CMAKE_SOURCE_DIR, then that should be a strict improvement because it will allow cache hits across multiple source directories, and multiple build directories, as long as build directories are not outside the source directories.
> 
> For the latter case, using different build directories will result in cache misses anyway, right?

It seems like that should be true but I haven't experimented or thought about this much and am just going off of the documentation. If cmake adds include paths inside the build directory, maybe for generated headers, or for headers are copied or linked to from there, I could see imagine CMAKE_BINARY_DIR being better to use than CMAKE_SOURCE_DIR in some cases. But naively I would expect CMAKE_SOURCE_DIR being better in most cases, so that led me to ask the question.

hebasto replied:
@ryanofsky 
> I only do the second not the first, because I use git worktrees...

Thanks for your feedback. Your concerns should be addressed in the recent [push](https://github.com/bitcoin/bitcoin/pull/30861#issuecomment-2351053745).

hebasto replied:
[Reworked](https://github.com/bitcoin/bitcoin/pull/30861#issuecomment-2596595610) to work equally well across both Git's wortktrees and build trees.

ryanofsky reviewed (approved):
Code review ACK 66b73bbf375ff96bf4a77539c9cf3fe52fbc15d5. This seems like a really helpful improvement, but did have some questions about it below.

hebasto commented on cmake/ccache.cmake:
```diff
@@ -24,7 +24,10 @@ if(NOT MSVC)
-    try_append_cxx_flags("-fdebug-prefix-map=A=B" TARGET core_interface SKIP_LINK
+    try_append_cxx_flags("-fdebug-prefix-map=A=B" SKIP_LINK
+      TARGET core_interface
+      # Propagate these flags, which apply to both C++ and C, to the secp256k1 subtree.
```
For now, only the `secp256k1` subtree uses its own build system. All other subtrees' targets consume the `core_interface`, which propagates these flags to them.

theuni replied:
Arguably the issue here is that we have debug symbols for secp at all.

hebasto replied:
@theuni 
> Arguably the issue here is that we have debug symbols for secp at all.

What are your suggestions?

hebasto replied:
This is no longer relevant for the recent [push](https://github.com/bitcoin/bitcoin/pull/30861#issuecomment-2596595610).

fanquake commented on src/CMakeLists.txt:
```diff
@@ -54,13 +54,17 @@ include(GetTargetInterface)
 # -fsanitize and related flags apply to both C++ and C,
 # so we can pass them down to libsecp256k1 as CFLAGS and LDFLAGS.
 get_target_interface(core_sanitizer_cxx_flags "" sanitize_interface COMPILE_OPTIONS)
-set(SECP256K1_APPEND_CFLAGS ${core_sanitizer_cxx_flags} CACHE STRING "" FORCE)
```
In 21bfce58c1cad31801e045c606bd8fb16a00e11d: Just generally, I'd say it's a bit less ideal, having secp256k1 flags/config getting split up into multiple (arbitrary) places. Especially if in future, we'll have to duplicate all this logic in all places, for all subtrees (when using their own cmake build systems).

hebasto replied:
This is no longer relevant for the recent [push](https://github.com/bitcoin/bitcoin/pull/30861#issuecomment-2596595610).

fanquake commented:
> What is your system and Ccache version?

That was on a macOS machine, with 4.10.2. Retested just now, and got the same: `Hits: 394 /  445 (88.54%)`.

fanquake commented:
Just tested the same on a Fedora box (ccache 4.10.2) and ended up with `Hits:  447 / 528 (84.66%)`.

hebasto commented:
> Just tested the same on a Fedora box (ccache 4.10.2) and ended up with `Hits: 447 / 528 (84.66%)`.

What if you `ccache --zero-stats` after configuring for a new build tree, just before building?

On Fedora, `ccache` masquerades as a compiler, so all compilation during configuration can affect the final hit rate.

l0rinc commented on cmake/module/TryAppendCXXFlags.cmake:
```diff
@@ -49,7 +43,7 @@ function(try_append_cxx_flags flags)
     "SKIP_LINK"                       # options
     "TARGET;VAR;SOURCE;RESULT_VAR"    # one_value_keywords
-    "IF_CHECK_PASSED;IF_CHECK_FAILED" # multi_value_keywords
+    "IF_CHECK_PASSED"                 # multi_value_keywords
```
nit: I've also removed `IF_CHECK_FAILED` in [`26cbe1e` (#30732)](https://github.com/bitcoin/bitcoin/pull/30732/commits/26cbe1e95d7960c6b4de19b77240b62f4f27d73b), consider readjusting the comment alignment as well

l0rinc commented on cmake/module/TryAppendCXXFlags.cmake:
```diff
@@ -49,7 +43,7 @@ function(try_append_cxx_flags flags)
     TACXXF                            # prefix
     "SKIP_LINK"                       # options
     "TARGET;VAR;SOURCE;RESULT_VAR"    # one_value_keywords
```
It seemed to me that `SOURCE` is also unused here, see: https://github.com/bitcoin/bitcoin/commit/031ac7f366263bacab152aa94a51025c582b1191#diff-d7f6f763fed50046f1c8a5dcca088fabb6b8c3662a55516f08fc0d0ba24829e3L45

hebasto replied:
This is no longer relevant for the recent [push](https://github.com/bitcoin/bitcoin/pull/30861#issuecomment-2596595610).

fanquake commented:
> What if you ccache --zero-stats after configuring for a new build tree, just before building?

In that case on Fedora it is `Hits: 446 / 446 (100.0%)`, with `Uncacheable calls: 12 / 458 ( 2.62%)`.

hebasto commented:
> > What if you ccache --zero-stats after configuring for a new build tree, just before building?
> 
> In that case on Fedora it is `Hits: 446 / 446 (100.0%)`, with `Uncacheable calls: 12 / 458 ( 2.62%)`.

If `ccache --show-stats --verbose` outputs:
```
Uncacheable calls:      12 / 458 ( 2.62%)
  Called for linking:   12 /  12 (100.0%)
```

this is expected on systems where `ccache` masquerades as a compiler.

hebasto commented:
The PR has been reworked to keep compatibility with git worktrees. See [this](https://github.com/bitcoin/bitcoin/pull/30861#discussion_r1752076990) discussion.

The PR description has been updated accordingly.

fanquake commented:
Seems like you missed some review comments?
https://github.com/bitcoin/bitcoin/pull/30861/files#r1752126150
https://github.com/bitcoin/bitcoin/pull/30861#discussion_r1752127638

Seems like https://github.com/bitcoin/bitcoin/pull/30861#discussion_r1752135299 is now also more relevant given #30905, so it'd be good to know how this is going to be handed going forward.

l0rinc commented on cmake/ccache.cmake:
```diff
@@ -17,15 +19,21 @@ if(NOT MSVC)
-      list(APPEND CMAKE_CXX_COMPILER_LAUNCHER ${CCACHE_EXECUTABLE})
+      foreach(lang IN ITEMS C CXX OBJCXX)
+        set(CMAKE_${lang}_COMPILER_LAUNCHER
+          ${CMAKE_COMMAND} -E env CCACHE_NOHASHDIR=1 ${CCACHE_EXECUTABLE}
```
Will this [affect coverage](https://github.com/ccache/ccache/discussions/268) or is that an old issue that doesn't affect us anymore?

hebasto replied:
I did not test this branch for code coverage reports.

stickies-v replied:
This does affect us wrt debug info. With this change, if you first compile in `tree1`, and then in `tree2`, inspecting the `.o` files in `tree2` will reference the source files in `tree1`. I don't know which, if any, use cases depend on that debug info being correct, but it's potentially annoying at least? 

Because env vars take precedence over config files, and because we're appending this, I don't think users would be able to undo this if they need correct debug info?

hebasto replied:
> This does affect us wrt debug info.

I don't think it affects us too much because:https://github.com/bitcoin/bitcoin/blob/fb0ada982a73687520c43b8fde480fa5d456f3e1/CMakeLists.txt#L456-L458

l0rinc commented on cmake/ccache.cmake:
```diff
@@ -17,15 +19,21 @@ if(NOT MSVC)
-      list(APPEND CMAKE_CXX_COMPILER_LAUNCHER ${CCACHE_EXECUTABLE})
+      foreach(lang IN ITEMS C CXX OBJCXX)
+        set(CMAKE_${lang}_COMPILER_LAUNCHER
+          ${CMAKE_COMMAND} -E env CCACHE_NOHASHDIR=1 ${CCACHE_EXECUTABLE}
```
> You can disable this option to get cache hits when compiling the same source code in different directories if you don’t mind that CWD in the debug info might be incorrect.

Does this affect us in any way?

hebasto replied:
If it happens, the project might be re-built with `-DWITH_CCACHE=OFF`.

l0rinc commented on cmake/ccache.cmake:
```diff
@@ -17,15 +19,21 @@ if(NOT MSVC)
-      list(APPEND CMAKE_CXX_COMPILER_LAUNCHER ${CCACHE_EXECUTABLE})
+      foreach(lang IN ITEMS C CXX OBJCXX)
+        set(CMAKE_${lang}_COMPILER_LAUNCHER
+          ${CMAKE_COMMAND} -E env CCACHE_NOHASHDIR=1 ${CCACHE_EXECUTABLE}
```
Note that `CCACHE_NOHASHDIR=1` was just removed in latest `master`'s CI config: https://github.com/bitcoin/bitcoin/pull/30869/files#diff-62587956f943bb2503db7bc6dd27d0d888074a1c0ecaab3f570ad611aff0f7bbL9

l0rinc reviewed (approved):
ACK 3f8afcefb53f18d735d9cd196df492d2d140c284

### Normal build on master:

`% ccache --clear --zero-stats`
`% cmake -B build1 && cmake --build build1 -j10`
`% ccache --show-stats`
```bash
Cacheable calls:    440 / 440 (100.0%)
  Hits:               0 / 440 ( 0.00%)
    Direct:           0
    Preprocessed:     0
  Misses:           440 / 440 (100.0%)
Local storage:
  Cache size (GiB): 0.2 / 5.0 ( 4.89%)
  Hits:               0 / 440 ( 0.00%)
  Misses:           440 / 440 (100.0%)
```
`% ccache --zero-stats`
`% cmake -B build2 && cmake --build build2 -j10`
`% ccache --show-stats`
```bash
Cacheable calls:    440 / 440 (100.0%)
  Hits:               0 / 440 ( 0.00%)
    Direct:           0
    Preprocessed:     0
  Misses:           440 / 440 (100.0%)
Local storage:
  Cache size (GiB): 0.5 / 5.0 ( 9.78%)
  Hits:               0 / 440 ( 0.00%)
  Misses:           440 / 440 (100.0%)
```

### Disabling `hash_dir` in ccache on `master`

`% ccache --set-config hash_dir=false`
`% ccache --zero-stats`
`% cmake -B build4 && cmake --build build4 -j10`
`% ccache --show-stats`
```bash
bitcoin % ccache --show-stats
Cacheable calls:    446 / 446 (100.0%)
  Hits:             324 / 446 (72.65%)
    Direct:          51 / 324 (15.74%)
    Preprocessed:   273 / 324 (84.26%)
  Misses:           122 / 446 (27.35%)
Local storage:
  Cache size (GiB): 1.1 / 5.0 (22.48%)
  Hits:             324 / 446 (72.65%)
  Misses:           122 / 446 (27.35%)
```
`% ccache --set-config hash_dir=true`

### Normal build on rebased branch:

`% ccache --clear --zero-stats`
`% cmake -B build1 && cmake --build build1 -j10`
`% ccache --show-stats`
```bash
Cacheable calls:    446 / 446 (100.0%)
  Hits:               0 / 446 ( 0.00%)
    Direct:           0
    Preprocessed:     0
  Misses:           446 / 446 (100.0%)
Local storage:
  Cache size (GiB): 0.2 / 5.0 ( 4.97%)
  Hits:               0 / 446 ( 0.00%)
```
`% ccache --zero-stats`
`% cmake -B build2 && cmake --build build2 -j10`
`% ccache --show-stats`
```bash
Cacheable calls:    446 / 446 (100.0%)
  Hits:             324 / 446 (72.65%)
    Direct:          51 / 324 (15.74%)
    Preprocessed:   273 / 324 (84.26%)
  Misses:           122 / 446 (27.35%)
Local storage:
  Cache size (GiB): 0.4 / 5.0 ( 7.74%)
  Hits:             324 / 446 (72.65%)
  Misses:           122 / 446 (27.35%)
```

### Debug build (directly after a normal build)

`% ccache --zero-stats`
`% cmake -B build_debug1 -DCMAKE_BUILD_TYPE=Debug && cmake --build build_debug1 -j10`
`% ccache --show-stats`
```bash
Cacheable calls:    446 / 446 (100.0%)
  Hits:               6 / 446 ( 1.35%)
    Direct:           0 /   6 ( 0.00%)
    Preprocessed:     6 /   6 (100.0%)
  Misses:           440 / 446 (98.65%)
Local storage:
  Cache size (GiB): 0.6 / 5.0 (12.41%)
  Hits:               6 / 446 ( 1.35%)
  Misses:           440 / 446 (98.65%)
```
`% ccache --zero-stats`
`% cmake -B build_debug2 -DCMAKE_BUILD_TYPE=Debug && cmake --build build_debug2 -j10`
`% ccache --show-stats`
```bash
Cacheable calls:    446 / 446 (100.0%)
  Hits:             324 / 446 (72.65%)
    Direct:          51 / 324 (15.74%)
    Preprocessed:   273 / 324 (84.26%)
  Misses:           122 / 446 (27.35%)
Local storage:
  Cache size (GiB): 0.8 / 5.0 (15.02%)
  Hits:             324 / 446 (72.65%)
  Misses:           122 / 446 (27.35%)
```

hebasto commented:
> Seems like you missed some review comments? https://github.com/bitcoin/bitcoin/pull/30861/files#r1752126150 [#30861 (comment)](https://github.com/bitcoin/bitcoin/pull/30861#discussion_r1752127638)
> 
> Seems like [#30861 (comment)](https://github.com/bitcoin/bitcoin/pull/30861#discussion_r1752135299) is now also more relevant given #30905, so it'd be good to know how this is going to be handed going forward.

Converting to a draft for now.

theuni commented:
Needs rebase after #31231

fanquake commented:
What is the status of this?

hebasto commented:
> What is the status of this?

Reworked. The PR description has been updated.

hebasto commented:
The CI failure is [unrelated](https://github.com/bitcoin/bitcoin/pull/31675).

hebasto commented on cmake/ccache.cmake:
```diff
@@ -17,8 +17,11 @@ if(NOT MSVC)
-      list(APPEND CMAKE_CXX_COMPILER_LAUNCHER ${CCACHE_EXECUTABLE})
+      foreach(lang IN ITEMS C CXX OBJCXX)
+        set(CMAKE_${lang}_COMPILER_LAUNCHER
+          ${CCACHE_EXECUTABLE} base_dir=${CMAKE_BINARY_DIR}
```
In the recent [push](https://github.com/bitcoin/bitcoin/pull/30861#issuecomment-2596595610), this branch relies on setting the `base_dir` option by the user because the required value depends on actual user's build environment.

davidgumberg commented:
Tested ACK, fixes worktree and build tree ccache hits.  https://github.com/bitcoin/bitcoin/commit/4d3f360e2af94f4ed46dc0943196d548da82003e


Note that as the PR description states, in the latest push, setting a `BASE_DIR` for `ccache` (as advised in the productivity notes) is essential for ccache to get hits in different directories.


<details>

<summary>Tested on gcc and clang on ubuntu 22.04</summary>

All on fresh ubuntu 22.04.1 vm images running running with qemu, in variations
I've omitted the three setup commands installing deps and git cloning.


```bash
# setup, ommitted in later examples
sudo apt upgrade -y
sudo apt install build-essential ccache clang cmake git libboost-dev libevent-dev libsqlite3-dev pkgconf -y
git clone https://github.com/bitcoin/bitcoin


export CC=clang CXX=clang++

mkdir -p ~/.config/ccache
echo  "base_dir = ${HOME}" > ~/.config/ccache/ccache.conf
cd bitcoin
git fetch origin pull/30861/head:30861 && git switch 30861
cmake -B build && cmake --build build -j $(nproc)
git worktree add ../other && cd ../other
cmake -B build && cmake --build build -j $(nproc)
```

```console
$ ccache --show-stats --verbose
Cacheable calls:     924 / 924 (100.0%)
  Hits:              462 / 924 (50.00%)
    Direct:          462 / 462 (100.0%)
    Preprocessed:      0 / 462 ( 0.00%)
  Misses:            462 / 924 (50.00%)
Successful lookups:
  Direct:            462 / 924 (50.00%)
  Preprocessed:        0 / 462 ( 0.00%)
Local storage:
  Cache size (GiB):  0.1 / 5.0 ( 2.42%)
  Files:             924
  Hits:              462 / 924 (50.00%)
  Misses:            462 / 924 (50.00%)
  Reads:            1848
  Writes:            924
```

## And buildtrees also work:

```bash
mkdir -p ~/.config/ccache
echo  "base_dir = ${HOME}" > ~/.config/ccache/ccache.conf
cd bitcoin
git fetch origin pull/30861/head:30861 && git switch 30861
cmake -B build && cmake --build build -j $(nproc)
cmake -B build2 && cmake --build build2 -j $(nproc)
```

```console
$ ccache --show-stats --verbose
Cacheable calls:     924 / 924 (100.0%)
  Hits:              462 / 924 (50.00%)
    Direct:          462 / 462 (100.0%)
    Preprocessed:      0 / 462 ( 0.00%)
  Misses:            462 / 924 (50.00%)
Successful lookups:
  Direct:            462 / 924 (50.00%)
  Preprocessed:        0 / 462 ( 0.00%)
Local storage:
  Cache size (GiB):  0.1 / 5.0 ( 2.42%)
  Files:             924
  Hits:              462 / 924 (50.00%)
  Misses:            462 / 924 (50.00%)
  Reads:            1848
  Writes:            924
````

### No basedir, pr branch, clang

<details> <summary>setup</summary>

```bash
export CC=clang CXX=clang++

# don't set a base dir in this run
# mkdir -p ~/.config/ccache
# echo  "base_dir = ${HOME}" > ~/.config/ccache/ccache.conf
cd bitcoin

git fetch origin pull/30861/head:30861 && git switch 30861
cmake -B build && cmake --build build -j $(nproc)
git worktree add ../other && cd ../other
cmake -B build && cmake --build build -j $(nproc)
```
</details>


```console 
$ ccache --show-stats --verbose
Cacheable calls:     924 / 924 (100.0%)
  Hits:                0 / 924 ( 0.00%)
    Direct:            0
    Preprocessed:      0
  Misses:            924 / 924 (100.0%)
Successful lookups:
  Direct:              0 / 924 ( 0.00%)
  Preprocessed:        0 / 924 ( 0.00%)
Local storage:
  Cache size (GiB):  0.2 / 5.0 ( 4.83%)
  Files:            1848
  Hits:                0 / 924 ( 0.00%)
  Misses:            924 / 924 (100.0%)
  Reads:            1848
  Writes:           1848
```


### basedir set, master, clang

<details> <summary>setup</summary>

```bash
export CC=clang CXX=clang++

mkdir -p ~/.config/ccache
echo  "base_dir = ${HOME}" > ~/.config/ccache/ccache.conf
cd bitcoin

# git fetch origin pull/30861/head:30861 && git switch 30861
cmake -B build && cmake --build build -j $(nproc)
git worktree add ../other && cd ../other
cmake -B build && cmake --build build -j $(nproc)
```
</details>

```console 
$ ccache --show-stats --verbose
Stats updated:      Tue Jan 28 23:44:43 2025
Cacheable calls:     924 / 924 (100.0%)
  Hits:                0 / 924 ( 0.00%)
    Direct:            0
    Preprocessed:      0
  Misses:            924 / 924 (100.0%)
Successful lookups:
  Direct:              0 / 924 ( 0.00%)
  Preprocessed:        0 / 924 ( 0.00%)
Local storage:
  Cache size (GiB):  0.2 / 5.0 ( 4.84%)
  Files:            1848
  Hits:                0 / 924 ( 0.00%)
  Misses:            924 / 924 (100.0%)
  Reads:            1848
  Writes:           1848
```

### basedir set, pr branch, gcc

<details> <summary>setup</summary>

```bash
export CC=gcc CXX=g++

mkdir -p ~/.config/ccache
echo  "base_dir = ${HOME}" > ~/.config/ccache/ccache.conf
cd bitcoin
git fetch origin pull/30861/head:30861 && git switch 30861
cmake -B build && cmake --build build -j $(nproc)
git worktree add ../other && cd ../other
cmake -B build && cmake --build build -j $(nproc)
```
</details>

```console
$ ccache --show-stats --verbose
Cacheable calls:     924 / 924 (100.0%)
  Hits:              462 / 924 (50.00%)
    Direct:          462 / 462 (100.0%)
    Preprocessed:      0 / 462 ( 0.00%)
  Misses:            462 / 924 (50.00%)
Successful lookups:
  Direct:            462 / 924 (50.00%)
  Preprocessed:        0 / 462 ( 0.00%)
Local storage:
  Cache size (GiB):  0.3 / 5.0 ( 6.55%)
  Files:             924
  Hits:              462 / 924 (50.00%)
  Misses:            462 / 924 (50.00%)
  Reads:            1848
  Writes:            924
```

### basedir set, master, gcc

<details> <summary>setup</summary>

```bash
export CC=gcc CXX=g++

mkdir -p ~/.config/ccache
echo  "base_dir = ${HOME}" > ~/.config/ccache/ccache.conf
cd bitcoin
# git fetch origin pull/30861/head:30861 && git switch 30861
cmake -B build && cmake --build build -j $(nproc)
git worktree add ../other && cd ../other
cmake -B build && cmake --build build -j $(nproc)
```
</details>

```console
$ ccache --show-stats --verbose
Cacheable calls:     924 / 924 (100.0%)
  Hits:              462 / 924 (50.00%)
    Direct:          462 / 462 (100.0%)
    Preprocessed:      0 / 462 ( 0.00%)
  Misses:            462 / 924 (50.00%)
Successful lookups:
  Direct:            462 / 924 (50.00%)
  Preprocessed:        0 / 462 ( 0.00%)
Local storage:
  Cache size (GiB):  0.3 / 5.0 ( 6.55%)
  Files:             924
  Hits:              462 / 924 (50.00%)
  Misses:            462 / 924 (50.00%)
  Reads:            1848
  Writes:            924
```


### No basedir, master, clang

<details> <summary>setup</summary>

```bash
export CC=clang CXX=clang++

# don't set a base dir in this run
# mkdir -p ~/.config/ccache
# echo  "base_dir = ${HOME}" > ~/.config/ccache/ccache.conf
cd bitcoin

# git fetch origin pull/30861/head:30861 && git switch 30861
cmake -B build && cmake --build build -j $(nproc)
git worktree add ../other && cd ../other
cmake -B build && cmake --build build -j $(nproc)
```

</details>

```console 
$ ccache --show-stats --verbose
Cacheable calls:     924 / 924 (100.0%)
  Hits:                0 / 924 ( 0.00%)
    Direct:            0
    Preprocessed:      0
  Misses:            924 / 924 (100.0%)
Successful lookups:
  Direct:              0 / 924 ( 0.00%)
  Preprocessed:        0 / 924 ( 0.00%)
Local storage:
  Cache size (GiB):  0.7 / 5.0 (13.14%)
  Files:            1848
  Hits:                0 / 924 ( 0.00%)
  Misses:            924 / 924 (100.0%)
  Reads:            1848
  Writes:           1848
```

TheCharlatan commented on cmake/ccache.cmake:
```diff
@@ -17,8 +19,11 @@ if(NOT MSVC)
-      list(APPEND CMAKE_C_COMPILER_LAUNCHER ${CCACHE_EXECUTABLE})
-      list(APPEND CMAKE_CXX_COMPILER_LAUNCHER ${CCACHE_EXECUTABLE})
+      foreach(lang IN ITEMS C CXX OBJCXX)
+        set(CMAKE_${lang}_COMPILER_LAUNCHER
```
Why is this now overriding the variable instead of appending?

hebasto replied:
Thanks!

I do not really have an explanation for this change...

We definitely should honour user-provided `CMAKE_<LANG>_COMPILER_LAUNCHER` environment variables if any.

The appending has been restored.

hebasto replied:
Reverted [back](https://github.com/bitcoin/bitcoin/pull/30861#issuecomment-2645380742) given this [comment](https://github.com/bitcoin/bitcoin/pull/30861#discussion_r1941284740).

sedited reviewed (approved):
ACK aeb3977db51792d8ad22b0a5ff8fc5ff20d69a00

davidgumberg commented:
utreACK https://github.com/bitcoin/bitcoin/commit/aeb3977db51792d8ad22b0a5ff8fc5ff20d69a00

stickies-v commented on cmake/ccache.cmake:
```diff
@@ -17,8 +19,9 @@ if(NOT MSVC)
-      list(APPEND CMAKE_C_COMPILER_LAUNCHER ${CCACHE_EXECUTABLE})
-      list(APPEND CMAKE_CXX_COMPILER_LAUNCHER ${CCACHE_EXECUTABLE})
+      foreach(lang IN ITEMS C CXX OBJCXX)
+        list(APPEND CMAKE_${lang}_COMPILER_LAUNCHER ${CMAKE_COMMAND} -E env CCACHE_NOHASHDIR=1 ${CCACHE_EXECUTABLE})
```
Perhaps this is obvious in CMake land, but I was confused by what `${CMAKE_COMMAND} -E env` is doing. I would have appreciated a docstring like the below:

>         # Use `cmake -E env` as a cross-platform way to set the CCACHE_NOHASHDIR environment variable, increasing
 >         # ccache hit rate e.g. in case of multiple build directories or git worktrees.

hebasto replied:
Thanks! The comment has been [added](https://github.com/bitcoin/bitcoin/pull/30861#issuecomment-2645380742).

stickies-v reviewed (commented):
Definitely in favour of increasing ccache effectiveness, but I'm still figuring out if this approach is optimal. Is my understanding correct that with this PR, all we're doing is forcing the `CCACHE_NOHASHDIR=1` environment variable, instead of letting the user set it?

While not unreasonable, it does seem to make it difficult for the user to undo this (without disabling ccache entirely, which also isn't ideal), and I'm not sure if this wouldn't break any workflows (see https://github.com/bitcoin/bitcoin/pull/30861#discussion_r1937579582)? If my understanding is correct, I think a viable alternative would be to mention this environment variable in our productivity notes?

willcl-ark commented on cmake/ccache.cmake:
```diff
@@ -2,7 +2,9 @@
-if(NOT MSVC)
+# The <LANG>_COMPILER_LAUNCHER target property, used to integrate
+# Ccache, is supported only by the Makefiles and Ninja generators.
+if(CMAKE_GENERATOR MATCHES "Make|Ninja")
```
We could also (or additionally) check if the user has set the launcher themselves by gating the entire following section with:

```cmake
 if(NOT DEFINED CMAKE_CXX_COMPILER_LAUNCHER)
 ```
 
 ...skipping all ccache setup if they have.
 
 This way we might be able to get the best of both "sane defaults" and "easy to override".

hebasto replied:
Thanks! [Reworked](https://github.com/bitcoin/bitcoin/pull/30861#issuecomment-2645380742) per your feedback.

hebasto committed:
cmake: Clarify ccache support for different generators

To enhance the user experience when using ccache, we utilize the CMake's
`<LANG>_COMPILER_LAUNCHER` target properties, which are only supported
by the Makefile and Ninja generators.

hebasto committed:
cmake: Enhance ccache performance across worktrees and build directories

To maximize performance, it is essential to follow the recommendations
in the Productivity Notes: set the `base_dir` ccache configuration
option or use the `CCACHE_BASEDIR` environment variable.

hebasto commented:
Thanks to all for the review!

Your feedback has been addressed.

hebasto commented on cmake/ccache.cmake:
```diff
@@ -2,7 +2,9 @@
 
-if(NOT MSVC)
+# The <LANG>_COMPILER_LAUNCHER target property, used to integrate
+# Ccache, is supported only by the Makefiles and Ninja generators.
```
Thanks! [Fixed](https://github.com/bitcoin/bitcoin/pull/30861#issuecomment-2645380742).

davidgumberg commented on cmake/ccache.cmake:
```diff
@@ -17,12 +19,23 @@ if(NOT MSVC)
-      list(APPEND CMAKE_CXX_COMPILER_LAUNCHER ${CCACHE_EXECUTABLE})
+      foreach(lang IN ITEMS C CXX OBJCXX)
+        if(DEFINED CMAKE_${lang}_COMPILER_LAUNCHER)
+          list(APPEND configure_warnings
+            "CMAKE_${lang}_COMPILER_LAUNCHER is already set. Skipping built-in ccache support for ${lang}."
+          )
```
Is making this a `configure_warning` compatible with #31665? Maybe it would be a benefit if CI complained when unintentionally skipping ccache.

hebasto replied:
Would you mind elaborating on your suggestion?

davidgumberg replied:
Maybe better expressed as: Is trying to both override the compiler launcher and make configure_warnings errors a use case that should be supported? If so, making this message a `configure_warning` is incompatible with #31665 since that makes `configure_warnings` fatal errors if `WCONFIGURE_ERROR=ON`.

If it is not a use case that should be supported, then making this message a configure_warning might be good together with #31665, since it would cause an error in CI if ccache ever got disabled because a compiler launcher was set, but maybe there isn't a realistic situation where something like that could happen unintentionally.

davidgumberg replied:
It's also fair if you just think #31665 is a bad approach and this is more evidence of that.

ryanofsky reviewed (approved):
Code review ACK 82a9f24f49b30c9462716cf0cbf207499be13edf since this seems like a reasonable change, but I'm confused about why it actually doing anything and agree with stickies concern https://github.com/bitcoin/bitcoin/pull/30861#pullrequestreview-2586373527 that forcing CCACHE_NOHASHDIR=1 might not be ideal, especially because according to documentation for [hash_dir](https://ccache.dev/manual/4.0.html#config_hash_dir) it sounds like if CWD is not hashed and `-fdebug-prefix-map` is not used, objects could be built with the wrong debug information.

But mainly I'm just confused why this change has any effect at all given that  [hash_dir](https://ccache.dev/manual/4.0.html#config_hash_dir) documentation says "The CWD will not be included in the hash if [base_dir](https://ccache.dev/manual/4.0.html#config_base_dir) is set (and matches the CWD) and the compiler option -fdebug-prefix-map is used." It seems like we are doing (or requiring) both of these things so I don't see why the hash_dir setting should change anything. The results posted https://github.com/bitcoin/bitcoin/pull/30861#issuecomment-2620364986 do seem to show it having an effect though, so think I must be misunderstanding something.

Also, I'm confused by why we need to require manually setting [base_dir](https://ccache.dev/manual/4.0.html#config_base_dir) instead of setting it automatically to CMAKE_SOURCE_DIR. I don't understand comment
https://github.com/bitcoin/bitcoin/pull/30861#discussion_r1757100406 about why this would fail to cache multiple build directories, since most of the time build directories will be under the source directory.

hebasto commented:
> Also, I'm confused by why we need to require manually setting [base_dir](https://ccache.dev/manual/4.0.html#config_base_dir) instead of setting it automatically to CMAKE_SOURCE_DIR.

Setting `base_dir` to `CMAKE_SOURCE_DIR` won't be helpful in cases when a worktree or a build directory resides outside of the repository root, for example, it is a sibling directory.

ryanofsky commented:
> Setting `base_dir` to `CMAKE_SOURCE_DIR` won't be helpful in cases when a worktree or a build directory resides outside of the repository root, for example, it is a sibling directory.

I don't understand why it wouldn't be helpful when a worktree is outside the repository root. If source directories are `/tree1/` and `/tree2`  and build directories are `/tree1/build` and `/tree2/build` I would expect source paths like `src/init.cpp` to be treated like `../src/init.cpp` in both cases and for the caching to work. Would also expect this to generalize regardless of the build directories are named as long as they are at the same level relative to the source directory. But, very possible I'm missing something since I haven't tested this.

I do understand that setting `base_dir` to `CMAKE_SOURCE_DIR`  will not work if developers are placing their build directories completely outside of their source directories, or at inconsistent depths within their source directories, but this seems much less common and not really a consideration if we are just deciding on a default base_dir value.

Separately, I'm still confused about why setting hash_dir has an effect (mentioned https://github.com/bitcoin/bitcoin/pull/30861#pullrequestreview-2630570206) so there are definitely things I'm not grasping here, but in any case this change seems at least safe and I am happy if it works in practice.

hebasto commented:
> I do understand that setting `base_dir` to `CMAKE_SOURCE_DIR` will not work if developers are placing their build directories completely outside of their source directories, or at inconsistent depths within their source directories, but this seems much less common and not really a consideration if we are just deciding on a default base_dir value.

Right. As for the PR author, it was not clear to me which cases are more common and which are less common, so I tried to take a more general approach.

Ultimately, the changes related to `base_dir` might be dropped, leaving it up to the user.

fanquake commented:
I'm going to re-add it, it'd be good to fix because there are a number of regressions here.

ryanofsky commented:
re: https://github.com/bitcoin/bitcoin/pull/30861#issuecomment-2682393121

> There is also a risk of adding even more regressions due to CCACHE_NOHASHDIR

@maflcko could you clarify this? Are you talking about the problem where objects might be built with the wrong paths in debug info mentioned https://github.com/bitcoin/bitcoin/pull/30861#pullrequestreview-2630570206 or a different problem? The debug path problem didn't seem like a real risk as long as we are specifying `-fdebug-prefix-map`.

---

re: https://github.com/bitcoin/bitcoin/pull/30861#issuecomment-2674076713

> I'm wondering in which scenarios a user might wish to disable the `-fdebug-prefix-map` flag?

I don't think there are any. Just mentioned this because `-fdebug-prefix-map` seems to be added conditionally in our build.

> > But mainly I'm just confused why this change has any effect at all given that [hash_dir](https://ccache.dev/manual/4.0.html#config_hash_dir) documentation says "The CWD will not be included in the hash if [base_dir](https://ccache.dev/manual/4.0.html#config_base_dir) is set (and matches the CWD) and the compiler option -fdebug-prefix-map is used." It seems like we are doing (or requiring) both of these things so I don't see why the hash_dir setting should change anything. The results posted [#30861 (comment)](https://github.com/bitcoin/bitcoin/pull/30861#issuecomment-2620364986) do seem to show it having an effect though, so think I must be misunderstanding something.
> 
> For the CWD to be omitted from hashing when a `base_dir` is set, it must _match_ the `base_dir`.

Right, the documentation I quoted says the base_dir must match the CWD.  But that just means the CWD needs "begin with base_dir" as stated in the [base_dir](https://ccache.dev/manual/4.0.html#config_base_dir) documentation. So it seems like regardless of how [hash_dir](https://ccache.dev/manual/4.0.html#config_hash_dir) is set, CWD should not be included in the hash anyway, so it is unclear why this PR has any effect. 

The change seems good and does seem to be helping according to https://github.com/bitcoin/bitcoin/pull/30861#issuecomment-2620364986 but I don't understand how it actually works if the documentation is accurate.

hebasto commented:
> > I'm wondering in which scenarios a user might wish to disable the `-fdebug-prefix-map` flag?
> 
> I don't think there are any. Just mentioned this because `-fdebug-prefix-map` seems to be added conditionally in our build.

Specifically, which condition are you referring to?

ryanofsky commented:
> > I don't think there are any. Just mentioned this because `-fdebug-prefix-map` seems to be added conditionally in our build.
> 
> Specifically, which condition are you referring to?

`-fdebug-prefix-map` is only added if `try_append_cxx_flags` succeeds and if the compilation target explicitly links against `core_interface`. It's reasonable to assume that these things are always true, but since correctness of the output when CCACHE_NOHASHDIR=1 is used seems to depend on them it would be good to document the assumptions where CCACHE_NOHASHDIR=1 is used. in general if code in one part of the build is relying on apparently unrelated code in a different part of the build that kind of thing is good to point out in case the assumptions are broken later.
